src/redux/selectors.js

```diff
@@ -0,0 +1,254 @@
+import { selectors as eventsSel } from './modules/events';
+import { selectors as tagsSel } from './modules/tags';
+import { selectors as sourcesSel } from './modules/sources';
+import { selectors as statsSel } from './modules/stats';
+import { selectors as playlistSel } from './modules/playlist';
+import { selectors as mySel } from './modules/my';
+import { selectors as recommendedSel } from './modules/recommended';
+import { selectors as playerSel } from './modules/player';
+import { selectors as assetsSel } from './modules/assets';
+import { selectors as chroniclesSel } from './modules/chronicles';
+import { selectors as searchSel } from './modules/search';
+import { selectors as listsSel } from './modules/lists';
+import { selectors as lessonsSel } from './modules/lessons';
+import { selectors as musicSel } from './modules/music';
+import { selectors as bookmarkFilterSel } from './modules/bookmarkFilter';
+import { selectors as simpleModeSel } from './modules/simpleMode';
+import { selectors as trimSel } from './modules/trim';
+import { selectors as fetchImageSel } from './modules/fetchImage';
+import { selectors as myNotesSel } from './modules/myNotes';
+
+// Select part of state (first parameter)
+export const getMDB     = state => state.mdb;
+const getAuth           = state => state.auth;
+const getEvents         = state => state.events;
+const getFilters        = state => state.filters;
+const getFiltersAside   = state => state.filtersAside;
+const getHome           = state => state.home;
+const getPublications   = state => state.publications;
+const getSettings       = state => state.settings;
+const getTags           = state => state.tags;
+const getSources        = state => state.sources;
+const getStats          = state => state.stats;
+const getPlaylist       = state => state.playlist;
+const getMy             = state => state.my;
+const getRecommended    = state => state.recommended;
+const getPlayer         = state => state.player;
+const getAssets         = state => state.assets;
+const getChronicles     = state => state.chronicles;
+const getSearch         = state => state.search;
+const getLists          = state => state.lists;
+const getLessons        = state => state.lessons;
+const getMusic          = state => state.music;
+const getBookmarkFilter = state => state.bookmarkFilter;
+const getSimpleMode     = state => state.simpleMode;
+const getTrim           = state => state.trim;
+const getFetchImage     = state => state.fetchImage;
+const getMyNotes        = state => state.myNotes;
+
+// Select additional parameters
+const _2 = (_, param) => param;
+const _3 = (_, __, param) => param;
+const _4 = (_, __, ___, param) => param;
+
+// auth
+export const authGetUserSelector = createSelector([getAuth], auth => authSel.getUser(auth));
+
+// settings
+export const settingsGetContentLanguagesSelector = createSelector(getSettings, _2, (s, skip) => settingsSel.getContentLanguages(s, skip));
```

Comment:
The magic here is createSelector?
If the result is cached, when will it be invalidated?

Reply:
I don't understand how this works.

```
const getContentLanguages = (state, skipFlags) => {
  if (state.urlLanguage.length && !skipFlags) {
    return state.urlLanguage;
  }

  if (state.showAllContent && !skipFlags) {
    return ALL_LANGUAGES;
  }

  return state.contentLanguages;
};
```

So what `createSelector` does? When will it know to invalidate the result and fetch getContentLanguages again?

Reply:
read here https://reselect.js.org/introduction/how-does-reselect-work#why-reselect-is-often-used-with-redux